### PR TITLE
[mlir][python] Clear PyOperations instead of invalidating them.

### PR DESCRIPTION
**Pull request by ingomueller-net**
`PyOperations` are Python-level handles to `Operation *` instances. When the latter are modified by C++, the former need to be invalidated. #69746 implements such invalidation mechanism by setting all `PyReferences` to `invalid`. However, that is not enough: they also need to be removed from the `liveOperations` map since other parts of the code (such as `PyOperation::createDetached`) assume that that map only contains valid refs.

This is required to actually solve the issue in #69730.

### CODE REVIEW

**Review by makslevental — APPROVED** [review]
Ya this is good. I think I was implicitly relying on a [`clearLiveOperations`](https://github.com/llvm/llvm-project/blob/f364a7a8b4915409ea32df4483320ad40a378242/mlir/lib/Bindings/Python/IRModule.h#L210) to happen after the pass but now I don't know why because as far I can tell (from my phone...) that function isn't actually called anywhere. Can you double check and if it's not then remove it.

File: mlir/include/mlir-c/IR.h [hunk 1]
```diff
@@ -705,12 +705,12 @@ typedef enum MlirWalkOrder {
 } MlirWalkOrder;
 
 /// Operation walker type. The handler is passed an (opaque) reference to an
-/// operation a pointer to a `userData`.
+/// operation and a pointer to a `userData`.
```
Comment (makslevental) [inline mlir/include/mlir-c/IR.h:708]: 😓

**Comment by ingomueller-net** [comment]
> Ya this is good. I think I was implicitly relying on a [`clearLiveOperations`](https://github.com/llvm/llvm-project/blob/f364a7a8b4915409ea32df4483320ad40a378242/mlir/lib/Bindings/Python/IRModule.h#L210) to happen after the pass but now I don't know why because as far I can tell (from my phone...) that function isn't actually called anywhere. Can you double check and if it's not then remove it.

Yeah, no, it's not called. It is also not precise enough: it would invalidate (and clear) *all* operations but with your original commit you are able to preserve the root op of the pass.

**Review by makslevental — APPROVED** [review]
Makes sense. Just curious - what are some other functions that might need to do this?

**Comment by ingomueller-net** [comment]
> Makes sense. Just curious - what are some other functions that might need to do this?

We have some out-of-tree extensions that run passes through their own binding functions. In the end, they also run passes, by they have a different entry point, where I want to call the invalidation logic from.

**Pull request merged**